### ISSUE TITLE
This program needs access to the screen

### ISSUE THREAD

**Issue by phiara** [facelessuser/Rummage#127]
I get the following error after running the command "rummage" in Terminal on my Mac:

```
Traceback (most recent call last):
  File "/Users/buenotech/anaconda/lib/python3.6/site-packages/rummage/lib/gui/notify/notify_osx.py", line 158, in setup
    assert(term_notify is not None and exists(term_notify))
AssertionError

This program needs access to the screen. Please run with a
Framework build of python, and only when you are logged in
on the main display of your Mac.
```

Is this a bug or am I doing something wrong?
I have installed the package with "pip install rummage".

Thanks.

**Comment by facelessuser** [comment]
How did you install Python? Did you do it with Homebrew or something like that?

**Comment by phiara** [comment]
I've installed Anaconda, I think that app installed Python.

**Comment by facelessuser** [comment]
Cool.  That's what I figured.  I think I ran into this once when attempting to run rummage from a virtual environment.  It appears to be a known issue with WxPython, but there are workarounds.

I haven't specifically used Anaconda though, so keep that in mind. I also never played with the solutions listed above myself. When I tried to run Rummage in a virtual machine and got this issue, and realized I just needed to run this directly from the framework build of Python (which I have installed) it worked fine.  I install my Python builds on Mac with the installers from https://www.python.org/.

So let's take a look at WxPython's statement on this: https://wiki.wxpython.org/wxPythonVirtualenvOnMac. 

Particularly pay attention to:

> All we need to do to fix this is to use the Python framework's python executable, but set it up in such a way that the virtual environment's library and site-packages will be used instead of those in the framework. It turns out that somebody has used Guido's Time Machine and implemented this feature for us in the form of the PYTHONHOME environment variable. By setting that to the virtualenv's location you can cause Python at runtime to use a different location as its install prefix, including where it looks for the site.py file which is where the virtualenv tool implements some of it's magic. I came up with this little script to help automate things for me:
> 
> ```
> #!/bin/bash
> 
> # what real Python executable to use
> PYVER=2.7
> PYTHON=/Library/Frameworks/Python.framework/Versions/$PYVER/bin/python$PYVER
> 
> # find the root of the virtualenv, it should be the parent of the dir this script is in
> ENV=`$PYTHON -c "import os; print os.path.abspath(os.path.join(os.path.dirname(\"$0\"), '..'))"`
> 
> # now run Python with the virtualenv set as Python's HOME
> export PYTHONHOME=$ENV 
> exec $PYTHON "$@"
> ```
> 
> All you need to do is copy this script into in a file in your environment's bin folder, (I called mine fwpy for FrameWork PYthon) and give it executable permissions. Then when you need to run some wxPython application in your virtual environment just use your copy of this script instead of "python" and you should be all set.

Let me know if one of the solutions on that page helps.  Since I haven't specifically played with these solutions, I didn't bother mentioning them in the documentation, but if this or one of the other suggested solutions works for you, I can update the docs to mention this bit of information.

**Comment by facelessuser** [comment]
Also maybe this discussion will help: https://github.com/conda/conda/issues/607.

**Comment by phiara** [comment]
Thank you very much for your support! I will look over these solutions and will come back with feedback.

**Comment by facelessuser** [comment]
No problem.  Also check this thread.  Here they specifically mention using pythonw when using Anaconda: https://groups.google.com/forum/#!topic/wxpython-users/ioH9uRx3Jh0.

If using miniconda they mention using the conda command to install Python.app to get ths pythonw command. So between those suggestions, hopefully you can get this working.

> I will look over these solutions and will come back with feedback.

This would be great!  I'd love to document stuff like this to help people with similar issues.

**Comment by phiara** [comment]
What I've managed to do is figure out what the problem is, but I still can't run rummage, yet.

Here are my findings:

I've installed python2.7 with **brew**, then followed the steps [here](https://wiki.wxpython.org/wxPythonVirtualenvOnMac).

On the step where you create a symbolic link I wrote the path to my Python2.7 **wxredirect.pth** file (**brew** will tell you where this is installed), which was called **wx.pth** on my install.

Running python with the custom script written on that page (fwpy), I don't get that error message in my first post. 

So the problem was that python was not finding and couldn't import the **wx** library.

Running rummage from that virtualenv is still throwing that error, but I think I'm close to solving it.
I'll write here if I figure it out.

**Comment by facelessuser** [comment]
Sounds great.  I appreciate the effort in figuring this out, I think it will help people in the future. It's possible those instructions on the wxPython wiki were outdated.

It's a shame that it is kind of difficult to get wxPython working with things like homebrew out of the box.  Part of the issue is probably due to the fact I am using the pip installable wxPython which is a pre-release `4.0.0a3`.  It appears brew does actually offer an installable wxpython, but is unfortunately on the last stable only `3.0.2.0`.

```
Faceless-MacBook-Pro:subclrschm facelessuser$ brew info wxpython
wxpython: stable 3.0.2.0 (bottled)
Python bindings for wxWidgets
https://www.wxwidgets.org/
Not installed
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/wxpython.rb
==> Dependencies
Required: wxmac ✔
```

I wanted to bring in PY3 support and make it easier to install which required the pre-release, but easier to install wxPython  4 version.  It is way easier to get up and running in Linux, Windows, and macOS (if using a framework build) when using the pip installable version.  But not so much on macOS when coupled with brew or Python installations installed in a similar manner.

There were so many changes going from wxPython 3 to wxPython 4 that I dropped support for wxPython 3.  It hadn't occurred to me all the people who'd be using things like brew to get Python instead of going the traditional route framework build route.  People may also not be too keen on using a non-stable wxpython build.  Ugh...I may need to provide an abstraction layer to bring back wxPython 3 support.  That sounds like a big pain ☹️ .

**Comment by facelessuser** [comment]
If you haven't figured out the virtualenv fix by this weekend, I'll probably spend some time over the weekend and see if I can throw together a tutorial to sidestep this framework build issue.

I don't know if I will actually have the motivation to abstract legacy wxPython 3.0.2.0 support in a new release.  More likely I would just create  a legacy branch or repo that people could manually install.  We'll have to see how motivated I am.  I may not get around to doing either; It's just so convenient moving forward with Phoenix 4.0.0a3+, and to be honest, it's no more buggy than wxPython has always been :slightly_smiling_face: .

**Comment by facelessuser** [comment]
@phiara So I was able to get this working in a virtualenv.

So WxPython 4 (Phoenix) doesn't really have a wx.pth as far as I can tell. Are you using wxPython 3.0.2.0?  The big change in WxPython4 moving forward is that it bundles it's libraries and such so you don't have a wx.pth.  I am afraid you might be using 3.0.2.0, but I switched over to using 4.0.0a3 which is installable via pip and supports Python2.7 and Python3.4+. Rummage no longer runs in 3.0.2.0 anymore.

All I needed to do is make my virtual environment use the actual binary framework python and then point at the root of my virtual python envrionment and its site-packages etc.

Here is the bash script which I called fwpy27 (frame work python 27).

```bash
#!/bin/bash

# what real Python executable to use
PYVER=2.7
PYTHON=/Library/Frameworks/Python.framework/Versions/$PYVER/bin/python$PYVER

# find the root of the virtualenv, it should be the parent of the dir this script is in
ENV=`$PYTHON -c "import os; print os.path.abspath(os.path.join(os.path.dirname(\"$0\"), '..'))"`

# now run Python with the virtualenv set as Python's HOME
export PYTHONHOME=$ENV
exec $PYTHON "$@"
```

This is me setting up my virtual environment, installing rummage into it, copying over the fwpy27 bash script, and lastly cd into the environment and run rummage.

```
Faceless-MacBook-Pro:Rummage facelessuser$ virtualenv 27-dist
New python executable in /Users/facelessuser/Desktop/Projects/Rummage/27-dist/bin/python
Installing setuptools, pip, wheel...done.

Faceless-MacBook-Pro:Rummage facelessuser$ 27-dist/bin/pip install rummage
Collecting rummage
  Using cached rummage-2.3-py2.py3-none-any.whl
Collecting wxpython>=4.0.0a3 (from rummage)
  Using cached wxPython-4.0.0a3-cp27-cp27m-macosx_10_6_intel.whl
Collecting backrefs>=1.0.1 (from rummage)
Collecting regex (from rummage)
Collecting gntp>=1.0.2 (from rummage)
  Using cached gntp-1.0.3-py2-none-any.whl
Collecting chardet>=3.0.4 (from rummage)
  Using cached chardet-3.0.4-py2.py3-none-any.whl
Collecting six (from wxpython>=4.0.0a3->rummage)
  Using cached six-1.10.0-py2.py3-none-any.whl
Installing collected packages: six, wxpython, backrefs, regex, gntp, chardet, rummage
Successfully installed backrefs-1.0.1 chardet-3.0.4 gntp-1.0.3 regex-2017.7.11 rummage-2.3 six-1.10.0 wxpython-4.0.0a3

Faceless-MacBook-Pro:Rummage facelessuser$ cp fwpy27 27-dist/bin/

Faceless-MacBook-Pro:Rummage facelessuser$ chmod +x 27-dist/bin/fwpy27

Faceless-MacBook-Pro:Rummage facelessuser$ cd 27-dist/

Faceless-MacBook-Pro:27-dist facelessuser$ bin/fwpy27 -m rummage
```

Now I'm not sure about brew yet, so I would have to see about that.  And if you **need** 3.0.2.0 support, I don't know if I am going to support that moving forward.  Getting this working in Python3 by switching to 4.0.0a3 is a big deal for me as I can drop using Python2 in my personal usage.  I will continue to support Python2 until end of life (2020), but I don't know if I plan on adding back in support for WxPython 3.0.2.0 (classic) along side WxPython 4.0.0a3 (phoenix).

I'll see if I can get this working under brew next.

**Comment by facelessuser** [comment]
I didn't have any issues with brew:

```
Faceless-MacBook-Pro:~ facelessuser$ brew install python
...a lot of install stuff

Faceless-MacBook-Pro:~ facelessuser$ /usr/local/Cellar/python/2.7.13_1/bin/pip install rummage
Collecting rummage
  Using cached rummage-2.3-py2.py3-none-any.whl
Collecting regex (from rummage)
Collecting backrefs>=1.0.1 (from rummage)
Collecting gntp>=1.0.2 (from rummage)
  Using cached gntp-1.0.3-py2-none-any.whl
Collecting chardet>=3.0.4 (from rummage)
  Using cached chardet-3.0.4-py2.py3-none-any.whl
Collecting wxpython>=4.0.0a3 (from rummage)
  Using cached wxPython-4.0.0a3-cp27-cp27m-macosx_10_6_intel.whl
Collecting six (from wxpython>=4.0.0a3->rummage)
  Using cached six-1.10.0-py2.py3-none-any.whl
Installing collected packages: regex, backrefs, gntp, chardet, six, wxpython, rummage
Successfully installed backrefs-1.0.1 chardet-3.0.4 gntp-1.0.3 regex-2017.7.11 rummage-2.3 six-1.10.0 wxpython-4.0.0a3

Faceless-MacBook-Pro:~ facelessuser$ /usr/local/Cellar/python/2.7.13_1/bin/python2 -m rummage
```

It worked fine.

**Comment by facelessuser** [comment]
Guess I'll go ahead and give Anaconda a try.

**Comment by facelessuser** [comment]
I also didn't have any issues with Anaconda.  I did a `pip install rummage` and used `pythonw -m rummage` (**not** `python -m rummage`, notice the `w`) and it worked.  No symlinking `wx.path` or anything crazy.  But remember, Rummage requires wxPython 4+, **not** wxPython 3.0.2.0.

**Comment by facelessuser** [comment]
Current notes and findings are now in the documentation: http://facelessuser.github.io/Rummage/installation/.

**Comment by phiara** [comment]
Ok, so with my virtualenv setup described in my last post it worked today. All I did was a OS restart. Weird.
I just run _fwpy -m rummage_.

By the way, your tool helped me move on with my project, so thank you for that :)

I guess you can close this thread now. Thank you.

**Comment by facelessuser** [comment]
> Ok, so with my virtualenv setup described in my last post it worked today.

Good to here.  After running through all those tests, I was pretty confident it should work.  It was a good exercise to run through.  Now I have an answer for anyone else trying to do this.  It seems that wxpython wiki is outdated, so I won't bother pointing people at that anymore 🙂 .

> By the way, your tool helped me move on with my project, so thank you for that :)

Glad to here it's helping others out!  Out of all the things I've written for myself, Rummage is probably the most useful thing I've created for my own personal workflow.  I use it all the time.

**Comment by ddrscott** [comment]
> I also didn't have any issues with Anaconda. I did a `pip install rummage` and used `pythonw -m rummage` (**not** `python -m rummage`, notice the `w`) and it worked. No symlinking `wx.path` or anything crazy. But remember, Rummage requires wxPython 4+, **not** wxPython 3.0.2.0.

`pythonw` worked for me

**Comment by klonos** [comment]
Sorry to revive this zombie-issue, but I am still having trouble (most likely due to my lack of understanding how `virtualenv` works I think).

I've installed python 3.x via `pyenv` (following advice [from this article](https://opensource.com/article/19/5/python-3-default-macos) - because nothing else I've been trying for months has worked):

```
pyenv install 3.7.3
python-build: use openssl from homebrew
python-build: use readline from homebrew
Downloading Python-3.7.3.tar.xz...
-> https://www.python.org/ftp/python/3.7.3/Python-3.7.3.tar.xz
Installing Python-3.7.3...
python-build: use readline from homebrew
python-build: use zlib from xcode sdk
Installed Python-3.7.3 to /Users/greg/.pyenv/versions/3.7.3
```

So here's what I've got:

```
pip -V
pip 19.1.1 from /Users/greg/.pyenv/versions/3.7.3/lib/python3.7/site-packages/pip (python 3.7)
```

```
python -V
Python 3.7.3
```

```
pyenv version
3.7.3 (set by /Users/greg/.pyenv/version)
```

```
brew info wxpython
wxpython: stable 4.0.6 (bottled)
Python bindings for wxWidgets
https://www.wxwidgets.org/
/usr/local/Cellar/wxpython/4.0.6 (1,130 files, 56.1MB) *
  Poured from bottle on 2019-06-15 at 05:23:44
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/wxpython.rb
==> Dependencies
Required: freetype ✔, jpeg ✔, libpng ✔, libtiff ✔, numpy ✔, python ✔
==> Analytics
install: 1,879 (30 days), 3,771 (90 days), 11,783 (365 days)
install_on_request: 967 (30 days), 2,094 (90 days), 8,285 (365 days)
build_error: 0 (30 days)
```

So far so good; but...

```
rummage
This program needs access to the screen. Please run with a
Framework build of python, and only when you are logged in
on the main display of your Mac.
```

```
python -m rummage
This program needs access to the screen. Please run with a
Framework build of python, and only when you are logged in
on the main display of your Mac.
```

```
pythonw -m rummage
/usr/local/opt/python@2/bin/python2.7: No module named rummage
```

^^ why is this still pointing to python 2.x??

```
fwpy -m rummage
bash: fwpy: command not found
```

So, what am I missing here? ...where am I to add this script?:

```
#!/bin/bash

# what real Python executable to use
PYVER=2.7
PYTHON=/Library/Frameworks/Python.framework/Versions/$PYVER/bin/python$PYVER

# find the root of the virtualenv, it should be the parent of the dir this script is in
ENV=`$PYTHON -c "import os; print os.path.abspath(os.path.join(os.path.dirname(\"$0\"), '..'))"`

# now run Python with the virtualenv set as Python's HOME
export PYTHONHOME=$ENV
exec $PYTHON "$@"
```

Noting that I know that I will mot likely need to:

- change `PYVER=2.7` to `PYVER=3.7`
- I have no `/Library/Frameworks/Python.framework` folder in my system

Any advice/pointers would be much appreciated.

**Comment by klonos** [comment]
@facelessuser I would like to point out that the documentation in https://facelessuser.github.io/Rummage/installation/#installing-in-virtual-environments-macos is missing this bit:

```
This program needs access to the screen. Please run with a
Framework build of python, and only when you are logged in
on the main display of your Mac.
```

...it just says "you may run into the following error:", but the error is missing. You may want to add it there so that people can find it if they are searching for it specifically. Thanks

**Comment by klonos** [comment]
...sorry for dumping terminal logs here, but this is all in the hope that these will help someone in the future:

So, I have tried wxpython:

```
brew reinstall wxpython
==> Reinstalling wxpython 
==> Installing dependencies for wxpython: python and numpy
==> Installing wxpython dependency: python
==> Downloading https://homebrew.bintray.com/bottles/python-3.7.3.mojave.bottle.tar.gz
Already downloaded: /Users/greg/Library/Caches/Homebrew/downloads/9d65c73313cc37259b0ee6b1369d5ef8a88dec8074b0cb74fb96aab7838e2d65--python-3.7.3.mojave.bottle.tar.gz
==> Pouring python-3.7.3.mojave.bottle.tar.gz
==> /usr/local/Cellar/python/3.7.3/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr/local/Cellar/python/3.7.3/bin --install-lib=/usr/local/lib/python3.7/site-packages --single-version-externally-managed --record=installed.txt
==> /usr/local/Cellar/python/3.7.3/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr/local/Cellar/python/3.7.3/bin --install-lib=/usr/local/lib/python3.7/site-packages --single-version-externally-managed --record=installed.txt
==> /usr/local/Cellar/python/3.7.3/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr/local/Cellar/python/3.7.3/bin --install-lib=/usr/local/lib/python3.7/site-packages --single-version-externally-managed --record=installed.txt
==> Caveats
Python has been installed as
  /usr/local/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /usr/local/opt/python/libexec/bin

If you need Homebrew's Python 2.7 run
  brew install python@2

You can install Python packages with
  pip3 install <package>
They will install into the site-package directory
  /usr/local/lib/python3.7/site-packages

See: https://docs.brew.sh/Homebrew-and-Python
==> Summary
🍺  /usr/local/Cellar/python/3.7.3: 3,863 files, 59.8MB
==> Installing wxpython dependency: numpy
==> Downloading https://homebrew.bintray.com/bottles/numpy-1.16.4.mojave.bottle.tar.gz
Already downloaded: /Users/greg/Library/Caches/Homebrew/downloads/d01cc1304ebca9df5d6092539fa9c3dd70e90aa534aceb498bd193e4468e0a47--numpy-1.16.4.mojave.bottle.tar.gz
==> Pouring numpy-1.16.4.mojave.bottle.tar.gz
sh: python: command not found
sh: python: command not found
==> Caveats
If you use system python (that comes - depending on the OS X version -
with older versions of numpy, scipy and matplotlib), you may need to
ensure that the brewed packages come earlier in Python's sys.path with:
  mkdir -p 
  echo 'import sys; sys.path.insert(1, "/usr/local/lib/python3.7/site-packages")' >> /homebrew.pth
==> Summary
🍺  /usr/local/Cellar/numpy/1.16.4: 1,000 files, 26.2MB
==> Installing wxpython
==> Downloading https://homebrew.bintray.com/bottles/wxpython-4.0.6.mojave.bottle.tar.gz
Already downloaded: /Users/greg/Library/Caches/Homebrew/downloads/75fc648a250ae5eed3e2d203d7280b3960d212d8d876dc0d082a80f6ef5626fc--wxpython-4.0.6.mojave.bottle.tar.gz
==> Pouring wxpython-4.0.6.mojave.bottle.tar.gz
Warning: wxpython dependency gcc was built with a different C++ standard
library (libstdc++ from clang). This may cause problems at runtime.
🍺  /usr/local/Cellar/wxpython/4.0.6: 1,130 files, 56.1MB
==> Caveats
==> python
Python has been installed as
  /usr/local/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /usr/local/opt/python/libexec/bin

If you need Homebrew's Python 2.7 run
  brew install python@2

You can install Python packages with
  pip3 install <package>
They will install into the site-package directory
  /usr/local/lib/python3.7/site-packages

See: https://docs.brew.sh/Homebrew-and-Python
==> numpy
sh: python: command not found
If you use system python (that comes - depending on the OS X version -
with older versions of numpy, scipy and matplotlib), you may need to
ensure that the brewed packages come earlier in Python's sys.path with:
  mkdir -p 
  echo 'import sys; sys.path.insert(1, "/usr/local/lib/python3.7/site-packages")' >> /homebrew.pth
```

Most noticeable things I will need to figure out seem to be:

- `sh: python: command not found`
-  ```
If you use system python (that comes - depending on the OS X version -
with older versions of numpy, scipy and matplotlib), you may need to
ensure that the brewed packages come earlier in Python's sys.path with:
  mkdir -p 
  echo 'import sys; sys.path.insert(1, "/usr/local/lib/python3.7/site-packages")' >> /homebrew.pth
```
- ```
Warning: wxpython dependency gcc was built with a different C++ standard
library (libstdc++ from clang). This may cause problems at runtime.
```

**Comment by facelessuser** [comment]
Is there a reason you need brew to install Python?  I saw in another issue you posted you had difficulty using the official installer, but you were referencing `python --version` which would grab Python 2 not Python 3 (`python3 --version`). The official installer is probably the best way to install it. When I have time, I can potentially go down this brew hole again, but if the official installer is a viable option for you, I would use that.

**Comment by klonos** [comment]
Hey @facelessuser ...I have responded to #346 with a bit of a background that explains why. I will not rehash all that here, to avoid noise.

> I can potentially go down this brew hole again...

No, I am a contributor to open source, and the last thing I would want is to waste your time...

> ... but if the official installer is a viable option for you, I would use that.

...I will take the time to make it an option. Please see my last response in #346 though. As this path always ends up in rummage 3.x installed on my system instead of the latest 4.x.

**Comment by facelessuser** [comment]
If your getting 3.x, you may be accidentally running your py 2.7 version. I'm able to install quite easily and run. I'm sure it is some misunderstanding that I can clear up, so if you give me the steps you are using, I can most likely let you know where things are going wrong. I already got it working in brew quite easily.

**Comment by klonos** [comment]
Problem sorted 👍 ...thanks so much for your time.

I will update #346 to keep things neat in a single place, so that people know that it is for python3 and rummage 4.x

**Comment by jeromel05** [comment]
Hello, 
Sorry to revive this old issue again. But I am stuck with the same error and I did not quite understand how you finally managed to solve your problem @klonos . Could you please explain the steps you took again? I have installed python, wxpython, python.app via conda not homebrew, so I do not get why wxpython would not be seen.

Any help is highly appreciated

Thank you

**Comment by facelessuser** [comment]
@jeromel05, can you mention how you are running it?

**Comment by facelessuser** [comment]
OS etc. may also be helpful to know.

**Comment by jeromel05** [comment]
Hello,

Yes sorry, here it is:
```
OS System Version:	macOS 10.14.3 (18D109) 
Packages version: 
wxpython                  4.0.7.post2              pypi_0    pypi                      
python                    3.7.6                h5c2c468_2    conda-forge
python.app                2                        py37_9    anaconda                                
ipykernel                 5.1.4            py37h5ca1d4c_0    conda-forge
ipython                   7.12.0           py37h5ca1d4c_0    conda-forge
conda                     4.8.2                    py37_0    anaconda
conda-build               3.18.11                  py37_0    conda-forge
conda-env                 2.6.0                         1    conda-forge
conda-package-handling    1.6.0            py37h0b31af3_1    conda-forge
anaconda                  custom                   py37_1  
anaconda-client           1.7.2                      py_0    conda-forge
anaconda-navigator        1.9.7                    py37_1  
notebook                  6.0.3                    py37_0    conda-forge
```

So I am running it either in a jupyter notebook like so:
```
import wx
wx.App()
```
And I have also tried adding the magic command `%gui wx` before

The error also shows up when I run it in terminal with ipython like so:
```
ipython
import wx
wx.App()
```

Thank you for taking the time to reply, I hope this information helps

**Comment by facelessuser** [comment]
I'm not quite understanding what you are trying to accomplish. Your above code is trying to create a wx app, but you are not calling rummage at all.

With that said, rummage, once in installed, is meant to be used as a command line app. If you ran it in a jupyter notebook, if imagine it just cause jupyter to hang until the app was closed.

You should just run it from the command line in your conda environment instead of trying to run it from a jupyter notebook.

**Comment by facelessuser** [comment]
Keep in mind that I don't use conda, but I have run rummage through it once in testing a while back.

**Comment by facelessuser** [comment]
Running rummage is documented here: https://facelessuser.github.io/Rummage/usage/#running. That is how it is expected to be run. Keep in mind that on Linux, I don't think there is such a thing as pythonw. In Linux you may have to just use `python -m rummage` or simply `rummage`.

**Comment by marekyggdrasil** [comment]
I had same problem on OSX with `pyenv`-managed Python environment. I described how I resolved it in [here](https://stackoverflow.com/questions/48531006/wxpython-this-program-needs-access-to-the-screen/66470363#66470363)

**Comment by facelessuser** [comment]
@marekyggdrasil Thanks for detailing your experience in this area. I know I don't run Rummage this way, but I know many people do attempt to. I may try to walk through this scenario sometime soon, if I can reproduce everything, along with the fix, I may try to document this better.

**Comment by vertok** [comment]
I've found the working solution for myself: 
https://stackoverflow.com/a/63502905/5870553
pip install -U -f https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-18.04 wxPython

**Comment by facelessuser** [comment]
@vertok, thanks for the info. I'm sure it will be helpful to others in the future.

**Comment by klonos** [comment]
I am running into this issue on Mac if I enable the "Secure Keyboard Entry" option in Terminal (which is a security requirement in the organization I work). Soon as you disable that option in Terminal, simply running `rummage` (installed via brew) works fine.

I hope that this information helps others running into this.

@facelessuser do you think that there's anything that can be done to make rummage run w/o any issues while having the "Secure Keyboard Entry" option turned on?

**Comment by facelessuser** [comment]
@klonos, I'm not exactly sure how to get around launching it from the terminal when "Secure Keyboard Entry" is enabled.  I've never enabled it. I just tried it, and it seems to disable Python's access to certain features which makes it runnable.

With that said, it can still be run other ways. You can add a quick action via Mac's Automator tool https://facelessuser.github.io/Rummage/extras/#file-manager-context-menu.  The documentation may be a little out of date as Mac seems to have changed the interface in recent versions, but:

Create a "Quick Action", set up as follows:

<img width="672" alt="Screenshot 2023-02-03 at 7 30 10 AM" src="https://user-images.githubusercontent.com/1055125/216628706-86204d16-1aaa-410a-b0ea-221be034b68c.png">

Save it. It should end up in `/Users/<username>/Library/Services/Rummage Here....`.

Once this is done, you should be able to  right click a file or folder, go to "Quick Actions", select `Rummage here...` (or whatever you named it).

<img width="531" alt="Screenshot 2023-02-03 at 7 34 39 AM" src="https://user-images.githubusercontent.com/1055125/216629813-43a4d70d-03d6-46a3-b57a-176f8b4cf57d.png">

If you use something like Alfred, you could create a workflow(s) to launch rummage from there.

<img width="575" alt="Screenshot 2023-02-03 at 7 39 56 AM" src="https://user-images.githubusercontent.com/1055125/216630917-3add1efb-0f8b-4b32-91d9-32e0de163707.png">

I'd have to research more to find a workaround to launch it directly in the terminal with "Secure Keyboard Entry" enabled.  Hopefully, some of this can help you work around the issue. Let me know if you have more questions.

**Comment by facelessuser** [comment]
I should probably consider making these available somehow so people could download them and alter them for their system...
